### PR TITLE
[JENKINS-36314] Tests failing after SECURITY-170

### PR DESCRIPTION
**Pull request by raul-arabaolaza**
[JENKINS-36314](https://issues.jenkins-ci.org/browse/JENKINS-36314?jql=project%20%3D%20JENKINS%20AND%20status%20in%20(Open%2C%20%22In%20Progress%22%2C%20Reopened)%20AND%20component%20%3D%20copyartifact-plugin%20AND%20text%20~%20%22SECURITY-170%22)

After SECURITY-170 parameterized jobs does not accept unknown build params as default, this PR explicitly sets the build params and copy permissions for the jobs used in the tests.

@reviewbybees

### CODE REVIEW

File: src/test/java/hudson/plugins/copyartifact/ParameterizedBuildSelectorTest.java [hunk 1]
```diff
@@ -27,12 +27,11 @@
 import static org.junit.Assert.*;
 
 import java.io.IOException;
+import java.util.Random;
 
+import hudson.model.*;
```
Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/ParameterizedBuildSelectorTest.java]: Please don't use wildcard imports.
Check the configuration of your IDE.

Comment (raul-arabaolaza) [inline src/test/java/hudson/plugins/copyartifact/ParameterizedBuildSelectorTest.java]: Sorry for that, fixed now

File: src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java [hunk 2]
```diff
@@ -346,8 +346,15 @@ public void testCopyToSlave() throws Exception {
...
         FreeStyleProject other = createArtifactProject(),
-                         p = createProject("$PROJSRC", null, "$BASE/*.txt", "$TARGET/bar",
+                         p = createProject("${PROJSRC}", null, "${BASE}/*.txt", "${TARGET}/bar",
                                            false, false, false, true);
+        ParameterDefinition paramDef = new StringParameterDefinition("PROJSRC",other.getName(), "");
+        ParameterDefinition paramDef2 = new StringParameterDefinition("BASE","","");
+        ParameterDefinition paramDef3 = new StringParameterDefinition("TARGET","foo","");
+        ParametersDefinitionProperty paramsDef = new ParametersDefinitionProperty(paramDef, paramDef2,paramDef3);
+        p.addProperty(paramsDef);
+        CopyArtifactPermissionProperty cpp=new CopyArtifactPermissionProperty("/"+p.getName());
```
Comment (jglick) [inline src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java]: Why does this need to be added now—why is it in this PR? Was JENKINS-20398 somehow related?

**Comment by jglick** [comment]
:bee:

Build timed out waiting for repository.jboss.org; we need to fix the CI job definition to use the Jenkins mirror exclusively.

File: src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java [hunk 3]
```diff
@@ -601,6 +608,11 @@ public void testJobRename() throws Exception {
     public void testSavedBuildSelector() throws Exception {
         FreeStyleProject other = createArtifactProject(),
                          p = rule.createFreeStyleProject();
+        ParameterDefinition paramDef = new StringParameterDefinition("FOO","foo");
+        ParametersDefinitionProperty paramsDef = new ParametersDefinitionProperty(paramDef);
+        other.addProperty(paramsDef);
+        CopyArtifactPermissionProperty cpp=new CopyArtifactPermissionProperty("/"+p.getName());
```
Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java]: I believe you don't need `/` as `CopyArtifactPernissionProperty` accepts relative names.
And you'd better to use `getFullName()` if you want to express with absolute names.

File: src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java [hunk 4]
```diff
@@ -619,6 +631,11 @@ public void testSavedBuildSelector() throws Exception {
     public void testSpecificBuildSelector() throws Exception {
         FreeStyleProject other = createArtifactProject(),
                          p = rule.createFreeStyleProject();
+        ParameterDefinition paramDef = new StringParameterDefinition("FOO","foo");
+        ParametersDefinitionProperty paramsDef = new ParametersDefinitionProperty(paramDef);
+        other.addProperty(paramsDef);
+        CopyArtifactPermissionProperty cpp=new CopyArtifactPermissionProperty("/"+p.getName());
```
Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java]: (about white spaces around `=`)
Though [the Jenkins project doesn't follow code styles strictly|https://wiki.jenkins-ci.org/display/JENKINS/Beginners+Guide+to+Contributing#BeginnersGuidetoContributing-CodeStyle], I highly recommend you to follow a some consistent code style.

File: src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java (hunk 2 above)
Comment (raul-arabaolaza) [inline src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java]: @jglick Because of SECURITY-170 jobs can not use parameters that are not explicitly declared, so if you run that test against jenkins 1.651.3 (last LTS) or jenkins over 2.3 (the ones with that patch applied) you get the following error and the PROJSRC, BASE and TARGET properties does not get expanded, resulting in the test failing

{noformat}
WARNING: Skipped parameter `PROJSRC` as it is undefined on `test1`. Set `-Dhudson.model.ParametersAction.keepUndefinedParameters`=true to allow undefined parameters to be injected as environment variables or `-Dhudson.model.ParametersAction.safeParameters=[comma-separated list]` to whitelist specific parameter names, even though it represents a security breach
Jun 30, 2016 9:41:12 AM hudson.model.ParametersAction filter
WARNING: Skipped parameter `BASE` as it is undefined on `test1`. Set `-Dhudson.model.ParametersAction.keepUndefinedParameters`=true to allow undefined parameters to be injected as environment variables or `-Dhudson.model.ParametersAction.safeParameters=[comma-separated list]` to whitelist specific parameter names, even though it represents a security breach
Jun 30, 2016 9:41:12 AM hudson.model.ParametersAction filter
WARNING: Skipped parameter `TARGET` as it is undefined on `test1`. Set `-Dhudson.model.ParametersAction.keepUndefinedParameters
Dhudson.model.ParametersAction.keepUndefinedParameters`=true to allow undefined parameters to be injected as environment variables or `-Dhudson.model.ParametersAction.safeParameters=[comma-separated list]` to whitelist specific parameter names, even though it represents a security breach
{noformat}

I honestly believe is better to explicitly add the needed params than use the hudson.model.ParametersAction.keepUndefinedParameters property

**Comment by raul-arabaolaza** [comment]
@ikedam Done!

File: src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java (hunk 2 above)
Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java]: -1 for adding unrelated changes, anyway.
And `CopyArtifactPermissionProperty` should be added only when it's actually required for tests.
(that is, Jenkins is secured and permissions to projects are configured)

File: src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java [hunk 5]
```diff
@@ -24,21 +24,29 @@
...
-import hudson.model.StringParameterValue;
-import hudson.model.FreeStyleBuild;
+import hudson.model.ParametersDefinitionProperty;
 import hudson.model.Result;
+import hudson.model.StringParameterDefinition;
+import hudson.model.StringParameterValue;
+import org.apache.commons.io.FileUtils;
+
+import hudson.Util;
+import hudson.maven.MavenModuleSet;
```
Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java]: I suppose this sorting of imports is performed by your IDE.
Could you revert this change?

Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java]: > I suppose this sorting of imports is performed by your IDE.
> Could you revert this change?

Because:
* It may cause other pull requests conflict.
* It gets difficult to track changes from git logs.

**Comment by raul-arabaolaza** [comment]
@ikedam Removed the unneeded copy permissions and rearranged the imports. Is now better??

PS: Sorry for wasting your time

File: src/test/java/hudson/plugins/copyartifact/ParameterizedBuildSelectorTest.java [hunk 6]
```diff
@@ -273,9 +284,13 @@ public void testVariableExpression() throws Exception {
         FreeStyleProject copiee = j.createFreeStyleProject();
         copiee.getBuildersList().add(new FileWriteBuilder("artifact.txt", "foobar"));
         copiee.getPublishersList().add(new ArtifactArchiver("artifact.txt"));
+        // Allow next project to copy from this
```
Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/ParameterizedBuildSelectorTest.java]: These lines should have been removed with removing `CopyArtifactPermissionProperty`.
You can review your whole changes in the "Files changed" tab in GitHub, and it should be more useful for you than review with `git diff`.

**Comment by raul-arabaolaza** [comment]
@ikedam Thanks for the tip, I found a lot of unneeded changes that I`ve reverted, what I have not reverted is the replace of some empty lines with hard tabs for empty lines and two existing * imports that I have changed to expanded ones. Would that be okay??

File: src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java [hunk 7]
```diff
@@ -551,6 +576,9 @@ public void testBackwardCompatibility() throws Exception {
         d.setGlobalUpstreamFilterStrategy(TriggeredBuildSelector.UpstreamFilterStrategy.UseOldest);
         
         FreeStyleProject upstream = j.createFreeStyleProject();
+        ParameterDefinition paramDef = new StringParameterDefinition("CONTENT","foo");
```
Comment (recena) [inline src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java]: `ParameterDefinition paramDef = new StringParameterDefinition("CONTENT", "foo");`

In this plugin, it is using this convention.

**Comment by recena** [comment]
👍

File: src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java (hunk 7 above)
Comment (raul-arabaolaza) [inline src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java]: @recena Fixed!!

**Comment by amuniz** [comment]
LGTM :bee:

**Comment by ikedam** [comment]
Thanks for updates.
Looks good to me.

**Comment by raul-arabaolaza** [comment]
Thanks @amuniz  @ikedam @recena  and @jglick for your help and comments

@reviewbybees done

**Comment by recena** [comment]
@raul-arabaolaza I've updated the [JIRA issue](https://issues.jenkins-ci.org/browse/JENKINS-36314) associated to this PR.
